blockchain/blockdao/filedao_v2_util.go

```diff
@@ -0,0 +1,246 @@
+		info := &block.Store{}
+		if err := info.Deserialize(v); err != nil {
+			return nil, err
+		}
+
+		// populate to staging buffer, if the block is in latest round
+		height := info.Block.Height()
+		if height > blockStoreTip {
+			buffer.Put(stagingKey(height, fd.bottomHeight), info)
+			fd.tipHeight = height
+		} else {
+			break
+		}
+	}
+	return buffer, nil
+}
+
+func (fd *fileDAOv2) putTipHashHeightMapping(blk *block.Block) error {
+	// write height <-> hash mapping
+	h := blk.HashBlock()
+	if err := addOneEntryToBatch(fd.hashStore, h[:], fd.batch); err != nil {
+		return err
+	}
+
+	// write hash <-> height mapping
+	heightValue := byteutil.Uint64ToBytesBigEndian(blk.Height())
+	fd.batch.Put(blockHashHeightMappingNS, hashKey(h), heightValue, "failed to put hash -> height mapping")
+
+	// update tip hash/height
+	if blk.Height() > fd.tipHeight {
+		fd.batch.Put(blockHashHeightMappingNS, topHeightKey, heightValue, "failed to put tip height")
+		fd.batch.Put(blockHashHeightMappingNS, topHashKey, h[:], "failed to put tip hash")
+	}
+	return nil
+}
+
+func (fd *fileDAOv2) putBlock(blk *block.Block) error {
+	blkInfo := &block.Store{
+		Block:    blk,
+		Receipts: blk.Receipts,
+	}
+	ser, err := blkInfo.Serialize()
+	if err != nil {
+		return err
+	}
+	blkBytes, err := compressDatabytes(ser, fd.compressBlock)
+	if err != nil {
+		return err
+	}
+
+	// add to staging buffer
+	index := stagingKey(blk.Height(), fd.bottomHeight)
+	full, err := fd.blkBuffer.Put(index, blkInfo)
+	if err != nil {
+		return err
+	}
+	if !full {
+		fd.batch.Put(stagingDataNS, byteutil.Uint64ToBytesBigEndian(index), blkBytes, "failed to put block")
+		return nil
+	}
+
+	// pack blockStoreBatchSize blocks together, write to block store
+	if ser, err = fd.blkBuffer.Serialize(); err != nil {
+		return err
+	}
+	if blkBytes, err = compressDatabytes(ser, fd.compressBlock); err != nil {
+		return err
+	}
+	return addOneEntryToBatch(fd.blkStore, blkBytes, fd.batch)
+}
+
+func (fd *fileDAOv2) putTransactionLog(blk *block.Block) error {
+	compressData := fd.compressBlock
+	sysLog := blk.TransactionLog()
+	if sysLog == nil {
+		sysLog = &block.BlkTransactionLog{}
+		compressData = false
+	}
+	logBytes, err := compressDatabytes(sysLog.Serialize(), compressData)
+	if err != nil {
+		return err
+	}
+	return addOneEntryToBatch(fd.sysStore, logBytes, fd.batch)
+}
+
+func addOneEntryToBatch(c db.CountingIndex, v []byte, b batch.KVStoreBatch) error {
+	if err := c.UseBatch(b); err != nil {
+		return err
+	}
+	if err := c.Add(v, true); err != nil {
+		return err
+	}
+	return c.Finalize()
+}
+
+func compressDatabytes(value []byte, comp bool) ([]byte, error) {
+	if value == nil {
+		return nil, ErrDataCorruption
+	}
+
+	// append 1 byte to end of bytestream to indicate the block is compressed or not
+	if !comp {
+		return append(value, _normal), nil
+	}
+
+	var err error
+	value, err = compress.CompV2(value)
+	if err != nil {
+		return nil, err
+	}
+	return append(value, _compressed), nil
```

Comment:
shall we just put if data is compressed with what compression schema in file header?

Reply:
fixed